taichi/transforms/constant_fold.cpp

```diff
@@ -14,49 +16,23 @@ class ConstantFold : public BasicStmtVisitor {
   ConstantFold() : BasicStmtVisitor() {
   }
 
-  struct JITEvaluatorId
-  {
-    int op;
-    DataType ret, lhs, rhs;
-    bool is_binary;
-
-    explicit operator JITEvaluatorIdType() const
-    {
-      // For a unique hash value, the number of UnaryOpTypes and BinaryOpTypes
-      // should be no more than 256, and the number of DataTypes should be no
-      // more than 128.
-      return (JITEvaluatorIdType)op | (JITEvaluatorIdType)ret << 8
-        | (JITEvaluatorIdType)lhs << 16 | (JITEvaluatorIdType)rhs << 24
-        | (JITEvaluatorIdType)is_binary << 31;
-    }
-
-    UnaryOpType unary_op() const
-    {
-      TI_ASSERT(!is_binary);
-      return (UnaryOpType) op;
-    }
-
-    BinaryOpType binary_op() const
+  static Kernel *get_jit_evaluator_kernel(JITEvaluatorId const &id) {
+    auto &cache = get_current_program().jit_evaluator_cache;
     {
-      TI_ASSERT(is_binary);
-      return (BinaryOpType) op;
+      std::lock_guard<std::mutex> _(
+          get_current_program().jit_evaluator_cache_mut);
+      auto it = cache.find(id);
+      if (it != cache.end())  // cached?
+        return it->second.get();
     }
```

Comment:
Why two separated lock guard? This will cause a situation that:
tid=1: find return -1
tid=2: find return -1
tid=3: save cache id 233
tid=4: save cache id 233
And id233 is stored twice.

Reply:
I guess it's probably better to store it twice rather than having one thread waiting for the other.

Reply:
> I guess it's probably better to store it twice rather than having one thread waiting for the other.

Right, there doesn't seem to be a perfect solution here...

Reply:
Okay, if the problem is fixed, then this is good enough. But could you add a comment for clarify this and a link like `// Discussion: xxx`?

Reply:
`Kernel::Kernel()` seems not supporting multi-threads, which is probably a reason why #2032 fails CI... So maybe we should use just one lock guard (or a reader-and-writer lock).